101-function-app-storage-private-endpoints/azuredeploy.parameters.json

```diff
@@ -2,5 +2,8 @@
   "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
   "contentVersion": "1.0.0.0",
   "parameters": {
+    "identityName": {
```

Comment:
you need to also output the resourceGroup name of the MSI in prereq.azuredeploy.json and then set the value here (the deploymentScript scenario isn't working)

Reply:
I believe I will need to put the roledef and role assignment into the main template in order for this to be able to pass the tests, unless I am missing another way to do this:
I see that the prereqs and main template are deployed to different resource groups.
The roledef would need to be assigned to the identity either at the subscription level (which isn't possible via the principal that is used in the tests) or in the scope of the main template's rg or resources (which isn't predeterminable via the tests). 

The reason I had put the role assignment in the prereqs template was to allow for the possibility that the deployer of the main template doesn't have Microsoft.Authorization/roleAssignments/write perms (so that someone with these perms could do the role assignment separately), but I can roll these up into the main template instead.

I will also make the "azPowerShellVersion": "5.5", "azCliVersion": "2.20.0" changes as well (I overlooked that comment in my last run).

Reply:
your design of putting the MSI as a prereq is spot-on... you just need to specify the resourceGroup parameter on the resourceId function that uses the MSI.  So your prereq template needs to output the name and the resourceGroup name - so in prereqs.json

outputs
 - name
 - resourcegroup

in azuredeploy.json
parameters
 - identityname
 - identityrg
...
resourceId(parameters(identityrg), Microsoft.ManagedIdentities/identities, parameters(identityname))

Reply:
So I did [these steps](https://github.com/Azure/azure-quickstart-templates/pull/8875#discussion_r592760854) in merge [b26d2a6](https://github.com/Azure/azure-quickstart-templates/pull/8875/commits/b26d2a671a0b14211a566135e6da628529f1eb92) , but Build [#48346](https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=48346&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=add333c3-6afd-5d0a-7233-d67f8d9bd855) failed due to the identity not having perms to do the necessary actions in the rg used by the main template (I verified this against the ARM backend logs via correlationId 4c07b8f1-8a29-4cfc-b93c-7a99c4a05f79 @ 2021-03-11 16:53 UTC ). I know that if I keep the MSI in the prereqs template and pass the identity name and identity rg to the main template, but move the role def & assignment to the main template akin to [this](https://github.com/Azure/azure-quickstart-templates/tree/master/201-machine-learning-dependencies-role-assignment) , it can succeed since the role def & assignment will be in the scope of the main rg. But I'm not seeing a way to have the role def & assignment be within the prereq but apply to the rg of the main template. Apologies in advance if I am overlooking something here.

Reply:
Ok, this is what it looks like you need:

MSI that has contributor access to the web/fn app.

If that's correct you'd do the following:

- create MSI
- Assign MSI the built-in role of contributor on the web that's created in the sample
- 
Since the MSI has a very constrained scope *and* the scope is a new web app, I would just create it in the main template, no prereq.  Your roleAssignment in the main template will look something like this:

```
    {
      "scope": "[concat('Microsoft.Web/sites/', parameters('appName')]",      
      "type": "Microsoft.Authorization/roleAssignments",
      "apiVersion": "2020-04-01-preview",
      "name": "[guid(resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', parameters('identityName')), variables('contributorRoleDefId'), parameters('appName'))]",
      "dependsOn": [
        "[resourceId('Microsoft.Web/sites', parameters('appName'))]"
      ],
      "properties": {
        "roleDefinitionId": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', variables('contributorRoleDefId'))]",
        "principalId": "[reference(resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', parameters('identityName')), '2018-11-30').principalId]",
        "principalType": "ServicePrincipal"
      }
    }
```

If that's still not working, just leave the commit in place so we can debug.